src/module_4/notebook.md

```diff
@@ -0,0 +1,2102 @@
+fig1.suptitle("Train metrics")
+
+fig2, ax2 = plt.subplots(1, 2, figsize=(12, 5))
+fig2.suptitle("Validation metrics")
+
+lrs = [
+    make_pipeline(StandardScaler(), LogisticRegression(penalty='l2', C=1e-6)),
+    make_pipeline(StandardScaler(), LogisticRegression(penalty='l1', C=1e-4, solver='saga')),
+    make_pipeline(RandomForestClassifier(n_estimators=100, max_depth=9)),
+    make_pipeline(GradientBoostingClassifier(n_estimators=100, max_depth=2, learning_rate=0.1)),
+]
+
+names = ["Ridge 1e-6", "Lasso 1e-4", "Random Forest", "Gradient Boosting"]
+for name, lr in zip(names, lrs):
+    lr.fit(X_train[train_cols], y_train)
+    train_proba = lr.predict_proba(X_train[train_cols])[:, 1]
+    plot_roc_and_pr_curves(name, y_pred=train_proba, y_test=y_train, figure=(fig1, ax1))
+    
+    val_proba = lr.predict_proba(X_val[train_cols])[:, 1]
+    plot_roc_and_pr_curves(name, y_pred=val_proba, y_test=y_val, threshold_target=0.05, figure=(fig2, ax2))
+    auc_scores_val = auc_score(y_val, val_proba)
+    roc_auc_scores_val.append(auc_scores_val[0])
+    pr_auc_score_val.append(auc_scores_val[1])
+    
+    
+plot_roc_and_pr_curves("Baseline", y_pred=val_df["global_popularity"], y_test=val_df[label_col], figure=(fig2, ax2))
+```
+
+
+    
+![png](notebook_files/notebook_38_0.png)
+    
+
+
+
+    
+![png](notebook_files/notebook_38_1.png)
+    
+
+
+
+```python
+models = {
+    "Model": ["Baseline"] + names,
+    "ROC AUC": [baseline_auc[0]] + roc_auc_scores_val,
+    "PR AUC": [baseline_auc[1]] + pr_auc_score_val
+    }
+final_results = pd.DataFrame(models)
+final_results
+```
+
+
+
+
+<div>
+<style scoped>
+    .dataframe tbody tr th:only-of-type {
+        vertical-align: middle;
+    }
+
+    .dataframe tbody tr th {
+        vertical-align: top;
+    }
+
+    .dataframe thead th {
+        text-align: right;
+    }
+</style>
+<table border="1" class="dataframe">
+  <thead>
+    <tr style="text-align: right;">
+      <th></th>
+      <th>Model</th>
+      <th>ROC AUC</th>
+      <th>PR AUC</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <th>0</th>
+      <td>Baseline</td>
+      <td>0.787093</td>
+      <td>0.066499</td>
+    </tr>
+    <tr>
+      <th>1</th>
+      <td>Ridge 1e-6</td>
+      <td>0.832559</td>
+      <td>0.160177</td>
+    </tr>
+    <tr>
+      <th>2</th>
+      <td>Lasso 1e-4</td>
+      <td>0.834434</td>
+      <td>0.148240</td>
+    </tr>
+    <tr>
+      <th>3</th>
+      <td>Random Forest</td>
+      <td>0.840997</td>
+      <td>0.198718</td>
+    </tr>
+    <tr>
+      <th>4</th>
+      <td>Gradient Boosting</td>
+      <td>0.844664</td>
+      <td>0.184307</td>
+    </tr>
+  </tbody>
+</table>
+</div>
+
+
+
+ Focusing on the AUC scores, we can see that RF is ahead of GBT, although not by much. If we check in the chart above, we can see that both ROC Curves are very similar. However, the thing changes when we focus on the PR Curve. With the RF ahead until a recall of ~0.4, from there, the two lines are similar.
```

Comment:
The idea would be know to evaluate the best model on test to see the non overfitted definitive metric